src/network/ClientConnectionManager.ts

```diff
@@ -125,47 +135,78 @@ export class ClientConnectionManager extends EventEmitter {
+    private readonly clusterService: ClusterService;
+    private readonly invocationService: InvocationService;
+
+    constructor(
+        client: ClientForClientConnectionManager,
+        clientName: string,
+        clientConfig: ClientConfig,
+        logger: ILogger,
+        partitionService: PartitionService,
+        serializationService: SerializationService,
+        lifecycleService: LifecycleService,
+        loadBalancer: LoadBalancer,
+        clusterFailoverService: ClusterFailoverService,
+        failoverConfig: ClientFailoverConfig,
+        clusterService: ClusterService,
+        invocationService: InvocationService
+    ) {
         super();
         this.client = client;
-        this.loadBalancer = client.getLoadBalancer();
-        this.labels = client.getConfig().clientLabels;
-        this.logger = this.client.getLoggingService().getLogger();
+        this.clientConfig = clientConfig;
+        this.clientName = clientName;
+        this.invocationService = invocationService;
+        this.serializationService = serializationService;
+        this.partitionService = partitionService;
+        this.clusterService = clusterService;
+        this.lifecycleService = lifecycleService;
+        this.loadBalancer = loadBalancer;
+        this.labels = this.clientConfig.clientLabels;
+        this.logger = logger;
         this.connectionTimeoutMillis = this.initConnectionTimeoutMillis();
-        this.heartbeatManager = new HeartbeatManager(client, this);
+        this.heartbeatManager = new HeartbeatManager(
+            clientConfig.properties,
+            this.logger,
+            this
+        );
         this.authenticationTimeout = this.heartbeatManager.getHeartbeatTimeout();
-        this.shuffleMemberList = client.getConfig().properties['hazelcast.client.shuffle.member.list'] as boolean;
-        this.smartRoutingEnabled = client.getConfig().network.smartRouting;
-        this.clusterDiscoveryService = client.getClusterFailoverService();
-        this.failoverConfigProvided = client.getFailoverConfig() != null;
-        this.waitStrategy = this.initWaitStrategy(client.getConfig() as ClientConfigImpl);
-        const connectionStrategyConfig = client.getConfig().connectionStrategy;
+        this.shuffleMemberList = this.clientConfig.properties['hazelcast.client.shuffle.member.list'] as boolean;
+        this.smartRoutingEnabled = this.clientConfig.network.smartRouting;
+        this.clusterDiscoveryService = clusterFailoverService;
+        this.failoverConfigProvided = failoverConfig != null;
+        this.waitStrategy = this.initWaitStrategy(this.clientConfig as ClientConfigImpl);
+        const connectionStrategyConfig = this.clientConfig.connectionStrategy;
         this.asyncStart = connectionStrategyConfig.asyncStart;
         this.reconnectMode = connectionStrategyConfig.reconnectMode;
     }
 
-    start(): Promise<void> {
+    start(invocationService: InvocationService): Promise<void> {
```

Comment:
On every invocation the InvocationService check if invocation allowed using a method on ClientConnectionManager instance. And that method depends on several things defined on ClientConnectionManager like clientState. Do not know what we can do. At least I can say that it would be a big change, isn't it?

Reply:
After some thinking I have a feeling that we should try to get simplify `ClientConnectionManager` as it seems to be one of the most complex classes in the client and does too much.

In order to do so we could try to extract `ClientConnectionManager`'s state which is consumed by other services (it seems to include active connections and client state) into a separate class. Then we could instantiate this object separately and pass it to all consumer services, as well as into `ClientConnectionManager` itself. As a side effect, this object will help us to remove some logic from the `ClientConnectionManager` which is good.

Here is a really draft version of that class. I don't like the fact that it combines active connections registry with client state (I feel that they are different things and should be kept separate), so please feel free to propose any improvements.
```ts
// the interface should describe read-only methods;
// it'll be consumed by all services except for ClientConnectionManager
// which needs to mutate the data
class ConnectionRegistryImpl implements ConnectionRegistry {

    private active = false;
    private activeConnections = new Map<string, ClientConnection>();
    private loadBalancer: LoadBalancer;
    private clientState = ClientState.INITIAL;
    private readonly smartRoutingEnabled: boolean;

    // constructor goes here
    constructor(/* ... */) {
        // ...
    }

    // not sure if that's a good name
    isActive(): boolean {
        return this.active;
    }

    isEmpty() {
        return this.activeConnections.size === 0;
    }

    getConnections(): ClientConnection[] {
        return Array.from(this.activeConnections.values());
    }

    getConnection(uuid: UUID): ClientConnection {
        return this.activeConnections.get(uuid.toString());
    }

    getRandomConnection(): ClientConnection {
        if (this.smartRoutingEnabled) {
            const member = this.loadBalancer.next();
            if (member != null) {
                const connection = this.getConnection(member.uuid);
                if (connection != null) {
                    return connection;
                }
            }
        }

        const iterator = this.activeConnections.values();
        const next = iterator.next();
        if (!next.done) {
            return next.value;
        } else {
            return null;
        }
    }

    forEachConnection(fn: (conn: ClientConnection) => void) {
        this.activeConnections.forEach(fn);
    }

    checkIfInvocationAllowed(): Error {
        // ...
    }

    // methods that mutate the state go here
    // ...
}
```

Reply:
What active property represents?

Reply:
Alright, I tried to implement the class and pushed the new commit. You can check jsdoc that I wrote for the functions as I am not sure of them.

Reply:
I have also implemented a different version in which we have a seperate class for ClientState. See [the change](https://github.com/srknzl/hazelcast-nodejs-client/commit/ee0835ad944265d584372df13e0265d501a48684) at seperate branch

Reply:
The code definitely got cleaner with `ConnectionRegistry` class. 👍

I'd prefer to keep both `ClientState` and the registry in a single object (at least, for now). Reviewed the update and left some ideas for further improvements.